File: go/internal/wire/catalog.go
package wire

import (
	"fmt"
	"net"
	"net/url"
	"slices"
	"strings"
)

const MaxCatalogServers = 32
const MaxSelectedServers = 4

// ServerEntry identifies one measurement authority, independently of its transport ports.
type ServerEntry struct {
	ID                string   `json:"id"`
	URL               string   `json:"url"`
	Name              string   `json:"name"`
	Location          string   `json:"location,omitempty"`
	AdditionalOrigins []string `json:"additionalOrigins,omitempty"`
}

type ServerCatalog struct {
	DefaultSelection []string      `json:"defaultSelection"`
	Servers          []ServerEntry `json:"servers"`
}

func SingletonCatalog() ServerCatalog {
	return ServerCatalog{DefaultSelection: []string{"self"},
		Servers: []ServerEntry{{ID: "self", URL: ".", Name: "graphite-meter"}}}
}

func (c ServerCatalog) Validate() error {
	if len(c.Servers) < 1 || len(c.Servers) > MaxCatalogServers || c.Servers[0].ID != "self" {
		return fmt.Errorf("catalogue requires self followed by at most %d additional servers", MaxCatalogServers-1)
	}
	ids, origins := map[string]bool{}, map[string]bool{}
	for _, entry := range c.Servers {
		if len(entry.ID) == 0 || len(entry.ID) > 64 || strings.ContainsFunc(entry.ID, func(r rune) bool {
			return !(r >= 'a' && r <= 'z' || r >= 'A' && r <= 'Z' || r >= '0' && r <= '9' ||
				r == '.' || r == '_' || r == '-')
		}) || len(entry.Name) > 256 || len(entry.Location) > 256 || !SafeText(entry.Name+entry.Location) {
			return fmt.Errorf("invalid catalogue server identity")
		}
		key, err := CanonicalOrigin(entry.URL)
		if entry.URL == "." && entry.ID != "self" || entry.URL != "." && err != nil {
			return fmt.Errorf("invalid catalogue origin for %q", entry.ID)
		}
		if ids[entry.ID] || origins[key] {
			return fmt.Errorf("duplicate catalogue server %q", entry.ID)
		}
		ids[entry.ID], origins[key] = true, true
		if len(entry.AdditionalOrigins) > 32 {
			return fmt.Errorf("too many additional origins for %q", entry.ID)
		}
		for _, raw := range entry.AdditionalOrigins {
			if _, err := CanonicalOrigin(raw); err != nil {
				return fmt.Errorf("invalid additional origin for %q", entry.ID)
			}
		}
	}
	return c.ValidateSelection(c.DefaultSelection)
}

func (c ServerCatalog) ValidateSelection(selected []string) error {
	if len(selected) < 1 || len(selected) > MaxSelectedServers {
		return fmt.Errorf("select one to %d servers", MaxSelectedServers)
	}
	seen := map[string]bool{}
	for _, id := range selected {
		if seen[id] || !slices.ContainsFunc(c.Servers, func(s ServerEntry) bool { return s.ID == id }) {
			return fmt.Errorf("unknown or repeated server %q", id)
		}
		seen[id] = true
	}
	return nil
}

func (c ServerCatalog) Resolve(base string) ServerCatalog {
	c.Servers = slices.Clone(c.Servers)
	for i, s := range c.Servers {
		raw := s.URL
		if raw == "." {
			raw = base
		}
		c.Servers[i].URL, _ = OriginKey(raw)
	}
	return c
}

// AllowsOrigin constrains discovery; it never grants credential access.
func (s ServerEntry) AllowsOrigin(raw string) bool {
	if raw == "." {
		return true
	}
	u, err := targetOrigin(raw)
	if err != nil {
		return false
	}
	base, err := url.Parse(s.URL)
	if err != nil {
		return false
	}
	if strings.EqualFold(u.Hostname(), base.Hostname()) {
		return true
	}
	return slices.ContainsFunc(s.AdditionalOrigins, func(allowed string) bool { return SameOrigin(raw, allowed) })
}

func (s ServerEntry) ValidateDiscovery(p Preflight) error {
	if err := p.Validate(); err != nil {
		return err
	}
	var origins []string
	for _, t := range p.Capabilities.ThroughputTargets {
		origins = append(origins, t.Origin)
	}
	for _, t := range p.Capabilities.LatencyTargets {
		origins = append(origins, t.Origin)
	}
	if !slices.ContainsFunc(origins, func(o string) bool { return !s.AllowsOrigin(o) }) {
		return nil
	}
	return fmt.Errorf("server %q advertised an unapproved target origin", s.ID)
}

// ConnectSources admits the configured hostname's transport ports, never subdomains.
func (c ServerCatalog) ConnectSources() []string {
	var out []string
	for _, s := range c.Servers {
		if s.URL == "." {
			continue
		}
		u, err := url.Parse(s.URL)
		if err != nil {
			continue
		}
		host := u.Hostname()
		if !strings.Contains(host, ":") {
			out = append(out, "http://"+host+":*", "https://"+host+":*", "ws://"+host+":*", "wss://"+host+":*")
		}
		for _, raw := range s.AdditionalOrigins {
			if BrowserConnectSourceSupported(raw) {
				out = append(out, raw, strings.Replace(raw, "http", "ws", 1))
			}
		}
	}
	return out
}

// BrowserConnectSourceSupported excludes IPv6 literals, which CSP host sources cannot express.
func BrowserConnectSourceSupported(raw string) bool {
	return !strings.Contains(raw, "://[")
}

// CanonicalOrigin is shared by catalogue decoders and authentication audiences.
func CanonicalOrigin(raw string) (string, error) {
	u, err := targetOrigin(raw)
	if err != nil || raw == "." {
		return "", fmt.Errorf("expected an absolute HTTP(S) origin")
	}
	if u.Port() == "0" || strings.ContainsAny(u.Hostname(), "*;") {
		return "", fmt.Errorf("invalid origin host or port")
	}
	if strings.Contains(u.Hostname(), ":") && net.ParseIP(u.Hostname()) == nil {
		return "", fmt.Errorf("invalid IPv6 origin")
	}
	key, _ := OriginKey(raw)
	return key, nil
}

// OriginKey is raw's scheme and host, lowercased and without a default port; ok is false if raw has neither.
func OriginKey(raw string) (string, bool) {
	u, err := url.Parse(raw)
	if err != nil || u.Scheme == "" || u.Hostname() == "" {
		return "", false
	}
	scheme, host, port := strings.ToLower(u.Scheme), strings.ToLower(u.Hostname()), u.Port()
	if scheme == "http" && port == "80" || scheme == "https" && port == "443" {
		port = ""
	}
	if port != "" {
		return scheme + "://" + net.JoinHostPort(host, port), true
	}
	if strings.Contains(host, ":") {
		// Hostname strips the brackets an IPv6 literal needs to be a valid authority.
		return scheme + "://[" + host + "]", true
	}
	return scheme + "://" + host, true
}

// SameOrigin reports whether a and b name one origin; text that is no origin matches nothing.
func SameOrigin(a, b string) bool {
	keyA, okA := OriginKey(a)
	keyB, okB := OriginKey(b)
	return okA && okB && keyA == keyB
}

File: rust/core/src/catalog.rs
//! Public measurement authorities and deterministic discovery boundaries.

use crate::{
    discovery::{DiscoveryError, Preflight},
    origin::{browser_connect_source_supported, canonical_origin, key, target_origin},
};
use serde::{Deserialize, Serialize};
use std::fmt;

pub const MAX_CATALOG_SERVERS: usize = 32;
pub const MAX_SELECTED_SERVERS: usize = 4;

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct ServerEntry {
    pub id: String,
    pub url: String,
    pub name: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    pub location: String,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub additional_origins: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServerCatalog {
    #[serde(default)]
    pub default_selection: Vec<String>,
    #[serde(default)]
    pub servers: Vec<ServerEntry>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CatalogError {
    InvalidServers,
    InvalidIdentity,
    DuplicateServer,
    InvalidOrigin,
    TooManyAdditionalOrigins,
    InvalidSelection,
}

impl fmt::Display for CatalogError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Self::InvalidServers => {
                "catalogue requires self followed by at most 31 additional servers"
            }
            Self::InvalidIdentity => "invalid catalogue server identity",
            Self::DuplicateServer => "duplicate catalogue server",
            Self::InvalidOrigin => "invalid catalogue origin",
            Self::TooManyAdditionalOrigins => "too many additional origins",
            Self::InvalidSelection => "select one to four distinct known servers",
        })
    }
}
impl std::error::Error for CatalogError {}

impl Default for ServerCatalog {
    fn default() -> Self {
        Self::singleton()
    }
}

impl ServerCatalog {
    pub fn singleton() -> Self {
        Self {
            default_selection: vec!["self".into()],
            servers: vec![ServerEntry {
                id: "self".into(),
                url: ".".into(),
                name: "graphite-meter".into(),
                ..ServerEntry::default()
            }],
        }
    }

    pub fn validate(&self) -> Result<(), CatalogError> {
        if self.servers.is_empty()
            || self.servers.len() > MAX_CATALOG_SERVERS
            || self.servers[0].id != "self"
        {
            return Err(CatalogError::InvalidServers);
        }
        for (index, entry) in self.servers.iter().enumerate() {
            let valid_id = !entry.id.is_empty()
                && entry.id.len() <= 64
                && entry
                    .id
                    .bytes()
                    .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'.' | b'_' | b'-'));
            let has_control_characters = entry
                .name
                .bytes()
                .chain(entry.location.bytes())
                .any(|byte| byte < 32 || byte == 127);
            if !valid_id
                || entry.name.len() > 256
                || entry.location.len() > 256
                || has_control_characters
            {
                return Err(CatalogError::InvalidIdentity);
            }
            let origin_key = key(&entry.url);
            if self.servers[..index]
                .iter()
                .any(|prior| prior.id == entry.id || key(&prior.url) == origin_key)
            {
                return Err(CatalogError::DuplicateServer);
            }
            if entry.url == "." {
                if entry.id != "self" {
                    return Err(CatalogError::InvalidOrigin);
                }
            } else if canonical_origin(&entry.url).is_err() {
                return Err(CatalogError::InvalidOrigin);
            }
            if entry.additional_origins.len() > 32 {
                return Err(CatalogError::TooManyAdditionalOrigins);
            }
            if entry
                .additional_origins
                .iter()
                .any(|raw| canonical_origin(raw).is_err())
            {
                return Err(CatalogError::InvalidOrigin);
            }
        }
        self.validate_selection(&self.default_selection)
    }

    pub fn validate_selection(&self, selected: &[String]) -> Result<(), CatalogError> {
        if selected.is_empty() || selected.len() > MAX_SELECTED_SERVERS {
            return Err(CatalogError::InvalidSelection);
        }
        for (index, id) in selected.iter().enumerate() {
            if selected[..index].contains(id) || !self.servers.iter().any(|entry| entry.id == *id) {
                return Err(CatalogError::InvalidSelection);
            }
        }
        Ok(())
    }

    pub fn resolve(&self, base: &str) -> Self {
        let mut resolved = self.clone();
        for entry in &mut resolved.servers {
            let origin = if entry.url == "." { base } else { &entry.url };
            entry.url = key(origin);
        }
        resolved
    }

    /// CSP transport ports for exact configured hosts, excluding IPv6 literals.
    pub fn connect_sources(&self) -> Vec<String> {
        let mut sources = Vec::new();
        for entry in &self.servers {
            let Ok(Some(origin)) = target_origin(&entry.url) else {
                continue;
            };
            if !origin.host.contains(':') {
                for scheme in ["http", "https", "ws", "wss"] {
                    sources.push(format!("{scheme}://{}:*", origin.host));
                }
            }
            for raw in &entry.additional_origins {
                if browser_connect_source_supported(raw) {
                    sources.push(raw.clone());
                    sources.push(raw.replacen("http", "ws", 1));
                }
            }
        }
        sources
    }
}

impl ServerEntry {
    pub fn validate_discovery(&self, preflight: &Preflight) -> Result<(), DiscoveryError> {
        preflight.validate()?;
        let throughput_origins = preflight
            .capabilities
            .throughput
            .iter()
            .map(|target| &target.base_url);
        let latency_origins = preflight
            .capabilities
            .latency
            .iter()
            .map(|target| &target.base_url);
        for origin in throughput_origins.chain(latency_origins) {
            if !self.allows_origin(origin) {
                return Err(DiscoveryError::UnapprovedOrigin);
            }
        }
        Ok(())
    }

    /// Constrains discovery only; never grants credential access.
    pub fn allows_origin(&self, raw: &str) -> bool {
        if raw == "." {
            return true;
        }
        let Ok(Some(origin)) = target_origin(raw) else {
            return false;
        };
        let base = match target_origin(&self.url) {
            Ok(base) => base,
            Err(_) => return false,
        };
        if base.is_some_and(|base| origin.host.eq_ignore_ascii_case(&base.host)) {
            return true;
        }
        self.additional_origins
            .iter()
            .any(|allowed| key(raw) == key(allowed))
    }
}
